src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Hover/RazorHoverEndpoint.cs

```diff
@@ -5,88 +5,95 @@
+            => DefaultHandleAsync(request, documentContext, codeDocument, sourceText, cancellationToken);
+
+        /// <inheritdoc/>
+        protected override Task<VSInternalHover?> RemapResponseAsync(VSInternalHover? response, RazorCodeDocument codeDocument, CancellationToken cancellationToken)
         {
-            if (request is null)
+            if (response is null)
             {
-                throw new ArgumentNullException(nameof(request));
+                throw new ArgumentNullException(nameof(response));
             }
 
-            var documentContext = await _documentContextFactory.TryCreateAsync(request.TextDocument.Uri, cancellationToken).ConfigureAwait(false);
-            if (documentContext is null)
+            if (response.Range is null)
             {
-                return null;
+                return Task.FromResult<VSInternalHover?>(response);
             }
 
-            var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken);
-            if (codeDocument.IsUnsupported())
+            if (_documentMappingService.TryMapToProjectedDocumentRange(codeDocument, response.Range, out var projectedRange))
             {
-                return null;
+                response.Range = projectedRange;
             }
 
-            var sourceText = await documentContext.GetSourceTextAsync(cancellationToken);
-            var linePosition = new LinePosition((int)request.Position.Line, (int)request.Position.Character);
-            var hostDocumentIndex = sourceText.Lines.GetPosition(linePosition);
-            var location = new SourceLocation(hostDocumentIndex, (int)request.Position.Line, (int)request.Position.Character);
-
-            var hoverItem = _hoverInfoService.GetHoverInfo(codeDocument, location, _clientCapabilities!);
-
-            _logger.LogTrace($"Found hover info items.");
-
-            return hoverItem;
+            return Task.FromResult<VSInternalHover?>(response);
         }
 
-        public RegistrationExtensionResult? GetRegistration(VSInternalClientCapabilities clientCapabilities)
+        private Task<VSInternalHover?> DefaultHandleAsync(VSHoverParamsBridge request, DocumentContext documentContext, RazorCodeDocument codeDocument, SourceText sourceText, CancellationToken _)
         {
-            const string AssociatedServerCapability = "hoverProvider";
-            _clientCapabilities = clientCapabilities;
-
-            var registrationOptions = new HoverOptions()
-            {
-                WorkDoneProgress = false,
-            };
+            var linePosition = new LinePosition(request.Position.Line, request.Position.Character);
+            var hostDocumentIndex = sourceText.Lines.GetPosition(linePosition);
+            var location = new SourceLocation(hostDocumentIndex, request.Position.Line, request.Position.Character);
+            var result = _hoverInfoService.GetHoverInfo(codeDocument, location, _clientCapabilities!);
```

Comment:
Ah given we only call this in the default (no single server) + language kind == Razor scenario I think this will break Component hovers. Try hovering over `<Counter>` in a Blazor app. I believe it'll return the language kind as HTML but we still want to provide hover information from this. It just turns out that HTML doesn't return any information for it. Similarly if you are to go into _Layout.cshtml and hover over one of the purple `<link` tags we'll need to have conflict resolution logic when both us + HTML return hover info

Reply:
Yeah, this was something that surprised me about rename, but made sense in the end. It's unlikely we actually want to ever check for "are we in razor", its more "try to handle this as though its razor, but if that doesn't work delegate".

It's also why I'm unsure about delegating in parallel

Reply:
Also scary if this doesn't cause any hover tests to fail 😛

Reply:
I fixed this, but keeping this open so I know to add a test

Reply:
Added